TestMultiSchemeAuthenticationHandler.java

```diff
@@ -0,0 +1,189 @@
+import static org.apache.hadoop.security.authentication.server.KerberosAuthenticationHandler.KEYTAB;
+import static org.apache.hadoop.security.authentication.server.KerberosAuthenticationHandler.NAME_RULES;
+import static org.apache.hadoop.security.authentication.server.LdapConstants.*;
+import static org.apache.hadoop.security.authentication.server.HttpConstants.*;
+
+import java.io.File;
+import java.util.Properties;
+
+import javax.servlet.http.HttpServletRequest;
+import javax.servlet.http.HttpServletResponse;
+
+import org.apache.commons.codec.binary.Base64;
+import org.apache.directory.server.annotations.CreateLdapServer;
+import org.apache.directory.server.annotations.CreateTransport;
+import org.apache.directory.server.core.annotations.ApplyLdifs;
+import org.apache.directory.server.core.annotations.ContextEntry;
+import org.apache.directory.server.core.annotations.CreateDS;
+import org.apache.directory.server.core.annotations.CreatePartition;
+import org.apache.directory.server.core.integ.AbstractLdapTestUnit;
+import org.apache.directory.server.core.integ.FrameworkRunner;
+import org.apache.hadoop.minikdc.KerberosSecurityTestcase;
+import org.apache.hadoop.security.authentication.KerberosTestUtils;
+import org.apache.hadoop.security.authentication.client.AuthenticationException;
+import org.junit.After;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.mockito.Mockito;
+
+/**
+ * This unit test verifies the functionality of "multi-scheme" auth handler.
+ */
+@RunWith(FrameworkRunner.class)
+@CreateLdapServer(
+    transports =
+      {
+        @CreateTransport(protocol = "LDAP", address = LDAP_SERVER_ADDR),
+      })
+@CreateDS(allowAnonAccess = true,
+          partitions = {
+            @CreatePartition(
+              name = "Test_Partition", suffix = LDAP_BASE_DN,
+              contextEntry = @ContextEntry(
+                  entryLdif = "dn: "+ LDAP_BASE_DN+ " \n"
+                              + "dc: example\n"
+                              + "objectClass: top\n"
+                              + "objectClass: domain\n\n"))})
+@ApplyLdifs({
+    "dn: uid=bjones," + LDAP_BASE_DN,
+    "cn: Bob Jones",
+    "sn: Jones",
+    "objectClass: inetOrgPerson",
+    "uid: bjones",
+    "userPassword: p@ssw0rd"})
+public class TestMultiSchemeAuthenticationHandler
+    extends AbstractLdapTestUnit {
+  private KerberosSecurityTestcase krbTest = new KerberosSecurityTestcase();
+  private MultiSchemeAuthenticationHandler handler;
+
+  @Before
+  public void setUp()  throws Exception {
+    krbTest.startMiniKdc();
+
+    // create keytab
+    File keytabFile = new File(KerberosTestUtils.getKeytabFile());
+    String clientPrinc = KerberosTestUtils.getClientPrincipal();
+    String serverPrinc = KerberosTestUtils.getServerPrincipal();
+    clientPrinc = clientPrinc.substring(0, clientPrinc.lastIndexOf("@"));
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer

If the lastIndexOf() search on line 90 does not find the substring and returns -1, the call to substring() on line 90 will throw IndexOutOfBoundsException.